# BMIN503/EPID600 Final Project
## WGCNA Application of Lung Adenocarcinoma on TCGA Data
In this study, RStudio was used to obtain correlation values and connectivity between differentially expressed genes. It is expected that TCGA LUAD samples will be divided into different immune infiltrating subtypes and associated gene modules and hub genes will be screened by weighted correlation network analysis (WGCNA). 

## Introduction
Lung cancer is still the leading cause of cancer death, and despite improvements in diagnosis, the death rate remains high. lung adenocarcinoma (LUAD) accounts for more than 40% of all lung cancer cases. Therefore, the treatment of LUAD is of great significance for improving the treatment quality of lung cancer patients. Natural genetic variation in the human genome is one reason for differences in individual responses to drugs. However, little research has been done on genetic differences between cancer patients. 
Therefore, in order to benefit patients, it is necessary to continue research on known targets or to find new targets associated with LUAD. In this study, we used RStudio to obtain correlation values and connectivity between differentially expressed genes. It is expected that TCGA LUAD samples will be divided into different immune infiltrating subtypes and associated gene modules and hub genes will be screened by weighted correlation network analysis (WGCNA). 
The correlation analysis of G protein-coupled receptors (GPCRs) and the screening of GPCR ligands have always been the focus of new drug development. In recent years, new studies have demonstrated the existence of newly discovered intracellular and extracellular interactions of GPCRs. Therefore, whether LUAD has unstudied GPCR related targets may be a novel direction. In similar studies, accurate identification of target-drug interactions is one of the key steps in the design of targeted drugs. However, screening drugs on an experimental scale is very expensive. Therefore, it is important to predict the interaction between substrate and drug pair directly from molecular sequence.

## Required Packages
- install "rtracklayer"
- install "edgeR"
- install "readr"
- library(data.table)
- library(dplyr)
- library(tidyr)

## Source Information
- The Cancer Genome Atlas (TCGA): https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga
- Perl: https://www.perl.org/
  - Copy and paste the following command into your Command Prompt and hit Enter to install this runtime along with the State Tool, our command line package manager:
  - powershell -Command "& $([scriptblock]::Create((New-Object Net.WebClient).DownloadString('https://platform.activestate.com/dl/cli/AS652/install.ps1'))) -activate-default Claire-WenyueZhao/Perl-5.34.0-Windows"
  - *The Perl used in this project is Perl 5.28.

##
1. To start, **fork** this BMIN503_Final_Project repository.
1. **Clone** the forked repository to your computer.
1. Modify the files provided, add your own, and **commit** changes to complete your final project.
1. **Push**/sync the changes up to your GitHub account.
1. Create a **pull request** on this, the original BMIN503_Final_Project, repository to turn in your final project.


Follow the instructions [here][forking] if you are unsure what the above steps mean.

DUE DATE FOR FINAL VERSION: 12/10/21 11:59PM. 


<!-- Links -->
[forking]: https://guides.github.com/activities/forking/

